boot/zephyr/decompression.c

```diff
@@ -1044,25 +1064,70 @@ int boot_copy_region_decompress(struct boot_loader_state *state, const struct fl
                                 goto finish;
                             }
 
-                            memcpy(&decomp_buf[processed_size], output_arm_thumb, current_size);
+                            memcpy(&decomp_buf[filter_writeback_pos], output_arm_thumb,
+                                   output_size_arm_thumb);
+                            filter_writeback_pos += output_size_arm_thumb;
                             processed_size += current_size;
                         }
-                    }
 
-                    rc = flash_area_write(fap_dst, (off_dst + hdr->ih_hdr_size + write_pos),
-                                          decomp_buf, sizeof(decomp_buf));
+                        if (excess_data_buffer_full == true)
+                        {
+                            /* Restore extra data removed from previous iteration to the write
+                             * buffer
+                             */
+                            memmove(&decomp_buf[DECOMP_BUF_EXTRA_SIZE], decomp_buf,
+                                    filter_writeback_pos);
+                            memcpy(decomp_buf, excess_data_buffer, DECOMP_BUF_EXTRA_SIZE);
+                            excess_data_buffer_full = false;
+                            filter_writeback_pos += DECOMP_BUF_EXTRA_SIZE;
+                        }
 
-                    if (rc != 0) {
-                        BOOT_LOG_ERR(
-                            "Flash write failed at offset: 0x%x, size: 0x%x, area: %d, rc: %d",
-                            (off_dst + hdr->ih_hdr_size + write_pos), sizeof(decomp_buf),
-                            fap_dst->fa_id, rc);
-                        rc = BOOT_EFLASH;
-                        goto finish;
-                    }
+                        if ((filter_writeback_pos % sizeof(uint32_t)) != 0)
+                        {
```

Comment:
OK, so the assumption is that filter_writeback_pos is aligned to at least two bytes? because modulo here can give 1 and 3 as well.

Reply:
The ARM thumb filter takes in data that is 4-byte aligned and outputs data that is 2-byte aligned so 1 and 3 are not possible

Reply:
^^ I was thought so.,